subprojects/execution/src/main/java/org/gradle/internal/execution/timeout/impl/DefaultTimeoutHandler.java

```diff
@@ -76,35 +87,73 @@ private static long warnIfNotStoppedFrequency() {
                     thread.interrupt();
-                    scheduledFuture = executor.schedule(this::warnIfNotStopped, warnIfNotStoppedFrequency(), TimeUnit.MILLISECONDS);
+                    logStacktraceTimer = Time.startCountdownTimer(slowStopLogStacktraceFrequency());
+                    scheduledFuture = executor.schedule(this::onAfterTimeoutCheck, postTimeoutCheckFrequency(), TimeUnit.MILLISECONDS);
                 }
             }
         }
 
-        private void warnIfNotStopped() {
+        private void onAfterTimeoutCheck() {
             synchronized (lock) {
                 if (!stopped) {
                     slowStop = true;
-                    doAsPartOfBuildOperation(() -> LOGGER.warn("Timed out {} has not yet stopped.", workUnitDescription.getDisplayName()));
-                    scheduledFuture = executor.schedule(this::warnIfNotStopped, warnIfNotStoppedFrequency(), TimeUnit.MILLISECONDS);
+                    doAsPartOfBuildOperation(() -> {
+                        LOGGER.warn("Timed out {} has not yet stopped.", workUnitDescription.getDisplayName());
+
+                        if (logStacktraceTimer.hasExpired()) {
+                            StackTraceElement[] currentStackTrace = thread.getStackTrace();
+                            if (currentStackTrace.length > 0 && !Arrays.equals(lastStacktrace, currentStackTrace)) {
+                                lastStacktrace = currentStackTrace;
+                                logStacktrace(currentStackTrace);
+                            }
+                            logStacktraceTimer.reset();
+                        }
+                    });
+
+                    thread.interrupt(); // interrupt again in case the work unit cleared the interrupt.
+
+                    scheduledFuture = executor.schedule(this::onAfterTimeoutCheck, postTimeoutCheckFrequency(), TimeUnit.MILLISECONDS);
+                }
+            }
+        }
+
+        private void logStacktrace(StackTraceElement[] currentStackTrace) {
+            if (LOGGER.isWarnEnabled()) {
+                // Assemble string this way so it is logged atomically, and with platform line endings
+                StringBuilder logMessageBuilder = new StringBuilder();
+                PrintWriter logMessageWriter = new PrintWriter(CharStreams.asWriter(logMessageBuilder));
+
+                logMessageWriter.print("Current stacktrace of timed out but not yet stopped ");
+                logMessageWriter.print(workUnitDescription.getDisplayName());
+                logMessageWriter.println(":");
+
+                for (StackTraceElement traceElement : currentStackTrace) {
+                    logMessageWriter.println("  at " + traceElement);
                 }
+
+                logMessageWriter.close();
```

Comment:
You could use try with resources to close the writer - or not close it at all.

Reply:
Went for try-with-resources.